langchain/agents/conversational_chat/output_parser.py

```diff
@@ -13,21 +13,35 @@ def get_format_instructions(self) -> str:
         return FORMAT_INSTRUCTIONS
 
     def parse(self, text: str) -> Union[AgentAction, AgentFinish]:
-        cleaned_output = text.strip()
-        if "```json" in cleaned_output:
-            _, cleaned_output = cleaned_output.split("```json")
-        if "```" in cleaned_output:
-            cleaned_output, _ = cleaned_output.split("```")
-        if cleaned_output.startswith("```json"):
-            cleaned_output = cleaned_output[len("```json") :]
-        if cleaned_output.startswith("```"):
-            cleaned_output = cleaned_output[len("```") :]
-        if cleaned_output.endswith("```"):
-            cleaned_output = cleaned_output[: -len("```")]
-        cleaned_output = cleaned_output.strip()
-        response = json.loads(cleaned_output)
-        action, action_input = response["action"], response["action_input"]
-        if action == "Final Answer":
-            return AgentFinish({"output": action_input}, text)
-        else:
-            return AgentAction(action, action_input, text)
+        class MessageJSONExtractor(json.JSONDecoder):
+            """
+            Custom JSON extractor that will extract the first JSON object in a string.
+            This is to handle the LLM returning more than just the JSON object.
+            """
+
+            def decode(self, s: str, _w: Any = None) -> Any:
+                for idx, char in enumerate(s):
+                    if char == "{":
+                        end_pos = self.raw_decode(s, idx)
+                        break
+                else:
+                    raise json.JSONDecodeError("No JSON object found", s, 0)
+
+                return end_pos[0]
+
+        try:
+            response = json.loads(text, cls=MessageJSONExtractor)
+            # only accept this result if it has the required fields
+            if all(field in response for field in ["action", "action_input"]):
+                if response["action"] == "Final Answer":
```

Comment:
sorry if this is a silly question. I'm still ramping up on agent part. Wondering if the action field could be a str enum?

btw customizing decoder is a brilliant idea! :D

Reply:
ChatGPT helped me write the json extractor heh.

Are you talking about the field's value? I'm not sure but I'm also new to this (and to python in general).

Reply:
yeah, field value. If "Final Answer" is one of the few predefined types, wondering if we can make a Enum for it for better type checking and easier revision in the future.

Reply:
Actually lately I've been wondering if "Final Answer" should be a tool and not hardcoded. I think it would help the LLM stay in a single mindset of always using tools.